Kernel/Arch/x86_64/PageDirectory.cpp

```diff
@@ -49,4 +56,92 @@ void activate_page_directory(PageDirectory const& pgd, Thread* current_thread)
     write_cr3(pgd.cr3());
 }
 
+UNMAP_AFTER_INIT NonnullLockRefPtr<PageDirectory> PageDirectory::must_create_kernel_page_directory()
+{
+    return adopt_lock_ref_if_nonnull(new (nothrow) PageDirectory).release_nonnull();
+}
+
+ErrorOr<NonnullLockRefPtr<PageDirectory>> PageDirectory::try_create_for_userspace()
+{
+    auto directory = TRY(adopt_nonnull_lock_ref_or_enomem(new (nothrow) PageDirectory));
+
+    directory->m_pml4t = TRY(MM.allocate_physical_page());
+
+    directory->m_directory_table = TRY(MM.allocate_physical_page());
+    auto kernel_pd_index = (kernel_mapping_base >> 30) & 0x1ffu;
+    for (size_t i = 0; i < kernel_pd_index; i++) {
+        directory->m_directory_pages[i] = TRY(MM.allocate_physical_page());
+    }
+
+    // Share the top 1 GiB of kernel-only mappings (>=kernel_mapping_base)
+    directory->m_directory_pages[kernel_pd_index] = MM.kernel_page_directory().m_directory_pages[kernel_pd_index];
+
+    {
+        InterruptDisabler disabler;
+        auto& table = *(PageDirectoryPointerTable*)MM.quickmap_page(*directory->m_pml4t);
+        table.raw[0] = (FlatPtr)directory->m_directory_table->paddr().as_ptr() | 7;
```

Comment:
i can suppose that this `7` is related to some kind of alignment… 🤔

yet, itʼs all unicorns for now! 😝

Reply:
The `7` is related to some flags of the x86_64 page directory entry. For the new aarch64 code you can see that I actually used the `TABLE_DESCRIPTOR` constant. The x86_64 code can use some cleanup too, but that can be done by someone else, or in a different PR.